src/App.js

```diff
@@ -0,0 +1,104 @@
+import CircularProgress from '@material-ui/core/CircularProgress'
+import Container from '@material-ui/core/Container'
+import makeStyles from '@material-ui/core/styles/makeStyles'
+import React, { useState, useEffect } from 'react'
+import NavBar from './components/NavBar'
+import ErrorSnackbar from './components/ErrorSnackbar'
+import { API_ROOT, get, params } from './api'
+import { styles } from './theme/styles'
+
+// Lazy load articles for faster page render time
+const ArticleGridListLazy = React.lazy(() => import('./components/ArticleGridList'))
+
+const useStyles = makeStyles(styles)
+
+const filterArticles = (articles, keyword) => {
+  // consider articles.results array:
+  // filter along title, byline, section
+  return articles.filter((article) => {
+    const lck = keyword.toLowerCase()
+    // TODO: simplify this mess, bruh
+    return (
+      article.title.toLowerCase().includes(lck) ||
+      article.byline.toLowerCase().includes(lck) ||
+      article.section.toLowerCase().includes(lck)
+    )
+  })
+}
+
+const App = () => {
+  const classes = useStyles()
+
+  // initialize keyword as empty string
+  const [
+    keyword,
+    setKeyword
+  ] = useState('')
+
+  // initialize articles state as empty array
+  // (to be lazily fetched via api HTTP get request)
+  const [
+    articles,
+    setArticles
+  ] = useState([])
+
+  const [
+    error,
+    setError
+  ] = useState({})
+
+  useEffect(
+    () => {
+      // TODO: clean up this fn
+      const getData = () =>
+        get(API_ROOT, params)
+          .then((response) => {
+            console.log(response)
+            if (response.ok) {
+              return response.json()
+            }
+            throw response
+          })
+          .catch((error) => {
+            if (error instanceof Error) {
+              return Promise.reject({ error })
+            }
+
+            return error.json().then((responseJson) => {
+              return Promise.reject({
+                error: new Error(`HTTP ${error.status} ${error.statusText}: ${responseJson.msg}`)
+              })
+            })
+          })
+          .then(
+            (responseData) => setArticles(responseData.results),
+            (error) => {
+              if (process.env.NODE_ENV === 'development') console.log(error)
+              setError({ error })
+            }
+          )
+      // call it!
+      getData()
+    },
+    [
+      setArticles,
+      setError
+    ]
+  )
+
+  let filteredArticles = filterArticles(articles, keyword)
+
+  return (
+    <div className='App'>
+      <NavBar siteTitle='Thrilling Articles' keyword={keyword} setKeyword={setKeyword} className={classes.navbar} />
+      <Container maxWidth='lg' className={classes.root}>
+      <ErrorSnackbar error={error} />
+        <React.Suspense fallback={<CircularProgress style={{ margin: 'auto', left: '50%', top: '50%' }} />}>
```

Comment:
[Suspense](https://reactjs.org/docs/concurrent-mode-suspense.html) is interesting. Not being very versed in React, I hadn't seen this before. In Angular you'd accomplish something similar using [Async Pipe](https://medium.com/front-end-weekly/several-ways-to-use-async-pipes-in-angular-fb0d080e8e12) to handle Promises or Observables. The one thing I don't get is how does Suspense know what data you're waiting for?

Reply:
I believe it waits for any props passed (in this case, `articles` and `keyword` to be defined) and render without error.

From the [documentation](https://reactjs.org/docs/concurrent-mode-suspense.html):
> Suspense is not a data fetching library. It’s a mechanism for data fetching libraries to communicate to React that the data a component is reading is not ready yet.